Adds a first batch of technical documentation (#53)
dedis/dela: docs/introduction.md

# Dela

Dela stands for DEDIS Ledger Architecture. It is both a set of abstractions and
an implementation of a distributed ledger architecture.

Dela has 2 main purposes:

- Provide a modular, global-purpose, and universal framework that describes a
  minimal and extended set of abstractions for a distributed ledger
  architecture.
- Provide multiple modules implementations that can be combined to run a
  distributed ledger.

With Dela you can:

- Learn the architecture of a distributed ledger
- Run your blockchain / distributed ledger
- Implement and test your new idea that will revolutionize the blockchain world
  by adding your new module's implementation to the Dela ecosystem

## Distributed ledger

A distributed ledger aims to solve the following problem: how can a set of
entities agree on a state without any central entity? In other words: how can a

> group of computers (or *nodes*) cooperate to maintain a database? This problem

Reviewer: ```suggestion
group of computers (or *nodes*) cooperates to maintain a database? This problem
```

Author: Actually this is right since I consider the group as a whole (https://brians.wsu.edu/2016/05/19/group-plural-vs-singular/)

Reviewer: Oh interesting !

A notorious application of a distributed ledger is Bitcoin, which uses a
blockchain data structure and a mining protocol (*Proof-of-Work*) to maintain a
consistent and decentralized list of coin transactions. This global list of
transactions (or *ledger*) is what allows the existence of a such decentralized
crypto-currency.

## Architecture

What is the best and minimal recipe for a distributed ledger architecture? We
asked ourself the same and came out with the following core ingredients:

- We first need an **overlay** module, which is responsible for transmitting
  messages from one entity to another one. If a group of human wants to talk and
  agree on a state, the overlay would be the common speaking language and the
  vocal cords that allow the persons to communicate and exchange pieces of
  information. With Dela, the overlay layer is called "Mino" for "Minimalist
  Network Overlay".
- Secondly, we need a **consensus** module, which describes how a set of
  entities can interact to agree on something. Most of the time it comes to the
  form of a zero/low trust interaction protocol. For example, in a teaching

>   class, a common consensus used is that the first to raise its hand can speak

Reviewer: ```suggestion
  class, a common consensus is that the first to raise its hand can speak
```

  and propose a solution. The consensus is the core part of a distributed ledger
  that makes the agreement on a state possible.

> - Then, we need a **ledger** module, which offers the mean to keep track of

Reviewer: ```suggestion
- Then, we need a **ledger** module, which keeps track of
```
